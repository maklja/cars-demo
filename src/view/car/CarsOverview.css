.cars-overview-grid .cars-overview-grid__item {
	padding: 5px; /* create gap between car images */
}
