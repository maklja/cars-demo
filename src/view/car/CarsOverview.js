import PropTypes from 'prop-types';
import React from 'react';
import 'bootstrap/dist/css/bootstrap-grid.css'; // use bootstrap 4 grid system to support responsive design

import { carPropTypes } from '../../props';
import './CarsOverview.css';
import CarPreview from './CarPreview';
import LoadingView from '../loading/LoadingView';

export const CarsOverview = ({
	cars,
	error,
	loading,
	onCarSelectionChange
}) => {
	let component = null;

	if (error != null) {
		// show error to the user
		component = (
			<div className="col-md-12">
				<div className="cars-overview-grid__message">
					Failed to fetch cars, please try again later.
				</div>
			</div>
		);
	}
	// check if request is still loading
	else if (loading) {
		// add delay of 2 sec, if request last less then 2 sec
		// there is no point in showing spinner to the user
		component = (
			<div className="col-md-12">
				<LoadingView delay={2000} text="Loading data, please wait..." />
			</div>
		);
	} else {
		// show cars
		component =
			cars.length > 0 ? (
				cars.map(curCar => (
					<div
						className="col-md-4 col-sm-6 col-xs-12 cars-overview-grid__item"
						key={curCar.id}
					>
						<CarPreview
							{...curCar}
							onSelectionChanged={onCarSelectionChange} // callback is invoked every time selection on element changes
						/>
					</div>
				))
			) : (
				<div className="col-md-12">
					<div className=" cars-overview-grid__message">
						No cars found.
					</div>
				</div>
			);
	}

	return (
		<div className="container-fluid">
			<div className="row cars-overview-grid">{component}</div>
		</div>
	);
};

CarsOverview.propTypes = {
	cars: PropTypes.arrayOf(PropTypes.shape(carPropTypes)).isRequired,
	loading: PropTypes.bool,
	error: PropTypes.instanceOf(Error),
	onCarSelectionChange: PropTypes.func
};

CarsOverview.defaultProps = {
	loading: false,
	error: null,
	onCarSelectionChange: () => {}
};

export default CarsOverview;
