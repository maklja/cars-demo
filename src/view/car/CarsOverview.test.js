import React from 'react';
import renderer from 'react-test-renderer';

import { CarsOverview } from './CarsOverview';

it('CarsOverview renders correctly with Mazda 3 and Mercedes Benz CLA', () => {
	const cars = [
		{
			image:
				'https://www.mazdausa.com/siteassets/vehicles/2017/m3s/trims/sport/17-m3s-sp-25d-snowflake-white.png',
			speed: 235,
			description:
				'The Fiat Punto is a supermini car produced by the Italian manufacturer Fiat since 1993.',
			name: 'Mazda 3',
			id: 2
		},
		{
			image: 'http://www.nriol.com/images/benz-cla.png',
			speed: 280,
			description:
				'The Fiat Punto is a supermini car produced by the Italian manufacturer Fiat since 1993.',
			name: 'Mercedes Benz CLA',
			id: 3
		}
	];

	const tree = renderer.create(<CarsOverview cars={cars} />).toJSON();
	expect(tree).toMatchSnapshot();
});

it('CarsOverview renders correctly with Mazda 3 and Mercedes Benz CLA, where Mazda 3 is selected', () => {
	const cars = [
		{
			image:
				'https://www.mazdausa.com/siteassets/vehicles/2017/m3s/trims/sport/17-m3s-sp-25d-snowflake-white.png',
			speed: 235,
			description:
				'The Fiat Punto is a supermini car produced by the Italian manufacturer Fiat since 1993.',
			name: 'Mazda 3',
			id: 2
		},
		{
			image: 'http://www.nriol.com/images/benz-cla.png',
			speed: 280,
			description:
				'The Fiat Punto is a supermini car produced by the Italian manufacturer Fiat since 1993.',
			name: 'Mercedes Benz CLA',
			id: 3
		}
	];
	const selectedCars = [2]; // select Mazda 3 on view

	const tree = renderer
		.create(<CarsOverview cars={cars} selectedCars={selectedCars} />)
		.toJSON();
	expect(tree).toMatchSnapshot();
});
