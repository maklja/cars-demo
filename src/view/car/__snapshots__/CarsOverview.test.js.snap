// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarsOverview renders correctly with Mazda 3 and Mercedes Benz CLA 1`] = `
<div
  className="container-fluid"
>
  <div
    className="row cars-overview-grid"
  >
    <div
      className="col-md-4 col-sm-6 col-xs-12 cars-overview-grid__item"
    >
      <div
        className="car-preview  "
        data-id={2}
        onClick={[Function]}
      >
        <div
          className="car-preview__card"
        >
          <div
            className="car-preview__front"
          >
            <img
              alt="Mazda 3"
              className="car-preview__image"
              src="https://www.mazdausa.com/siteassets/vehicles/2017/m3s/trims/sport/17-m3s-sp-25d-snowflake-white.png"
            />
          </div>
          <div
            className="car-preview__back"
          >
            <div
              className="car-preview__overlay"
            />
            <div
              className="car-preview__description"
            >
              <div
                className="car-preview__description-text"
              >
                The Fiat Punto is a supermini car produced by the Italian manufacturer Fiat since 1993.
              </div>
              <div>
                Top speed: 
                235
                km/h
              </div>
            </div>
            <img
              alt="Mazda 3"
              className="car-preview__image"
              src="https://www.mazdausa.com/siteassets/vehicles/2017/m3s/trims/sport/17-m3s-sp-25d-snowflake-white.png"
            />
          </div>
        </div>
        <div
          className="car-preview__title"
        >
          Mazda 3
        </div>
      </div>
    </div>
    <div
      className="col-md-4 col-sm-6 col-xs-12 cars-overview-grid__item"
    >
      <div
        className="car-preview  "
        data-id={3}
        onClick={[Function]}
      >
        <div
          className="car-preview__card"
        >
          <div
            className="car-preview__front"
          >
            <img
              alt="Mercedes Benz CLA"
              className="car-preview__image"
              src="http://www.nriol.com/images/benz-cla.png"
            />
          </div>
          <div
            className="car-preview__back"
          >
            <div
              className="car-preview__overlay"
            />
            <div
              className="car-preview__description"
            >
              <div
                className="car-preview__description-text"
              >
                The Fiat Punto is a supermini car produced by the Italian manufacturer Fiat since 1993.
              </div>
              <div>
                Top speed: 
                280
                km/h
              </div>
            </div>
            <img
              alt="Mercedes Benz CLA"
              className="car-preview__image"
              src="http://www.nriol.com/images/benz-cla.png"
            />
          </div>
        </div>
        <div
          className="car-preview__title"
        >
          Mercedes Benz CLA
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`CarsOverview renders correctly with Mazda 3 and Mercedes Benz CLA, where Mazda 3 is selected 1`] = `
<div
  className="container-fluid"
>
  <div
    className="row cars-overview-grid"
  >
    <div
      className="col-md-4 col-sm-6 col-xs-12 cars-overview-grid__item"
    >
      <div
        className="car-preview car-preview--selected "
        data-id={2}
        onClick={[Function]}
      >
        <div
          className="car-preview__card"
        >
          <div
            className="car-preview__front"
          >
            <img
              alt="Mazda 3"
              className="car-preview__image"
              src="https://www.mazdausa.com/siteassets/vehicles/2017/m3s/trims/sport/17-m3s-sp-25d-snowflake-white.png"
            />
          </div>
          <div
            className="car-preview__back"
          >
            <div
              className="car-preview__overlay"
            />
            <div
              className="car-preview__description"
            >
              <div
                className="car-preview__description-text"
              >
                The Fiat Punto is a supermini car produced by the Italian manufacturer Fiat since 1993.
              </div>
              <div>
                Top speed: 
                235
                km/h
              </div>
            </div>
            <img
              alt="Mazda 3"
              className="car-preview__image"
              src="https://www.mazdausa.com/siteassets/vehicles/2017/m3s/trims/sport/17-m3s-sp-25d-snowflake-white.png"
            />
          </div>
        </div>
        <div
          className="car-preview__title"
        >
          Mazda 3
        </div>
      </div>
    </div>
    <div
      className="col-md-4 col-sm-6 col-xs-12 cars-overview-grid__item"
    >
      <div
        className="car-preview  "
        data-id={3}
        onClick={[Function]}
      >
        <div
          className="car-preview__card"
        >
          <div
            className="car-preview__front"
          >
            <img
              alt="Mercedes Benz CLA"
              className="car-preview__image"
              src="http://www.nriol.com/images/benz-cla.png"
            />
          </div>
          <div
            className="car-preview__back"
          >
            <div
              className="car-preview__overlay"
            />
            <div
              className="car-preview__description"
            >
              <div
                className="car-preview__description-text"
              >
                The Fiat Punto is a supermini car produced by the Italian manufacturer Fiat since 1993.
              </div>
              <div>
                Top speed: 
                280
                km/h
              </div>
            </div>
            <img
              alt="Mercedes Benz CLA"
              className="car-preview__image"
              src="http://www.nriol.com/images/benz-cla.png"
            />
          </div>
        </div>
        <div
          className="car-preview__title"
        >
          Mercedes Benz CLA
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`CarsOverview renders correctly with error message 1`] = `
<div
  className="container-fluid"
>
  <div
    className="row cars-overview-grid"
  >
    <div
      className="col-md-12"
    >
      <div
        className="cars-overview-grid__message"
      >
        Failed to fetch cars, please try again later.
      </div>
    </div>
  </div>
</div>
`;

exports[`CarsOverview renders correctly with loading message 1`] = `
<div
  className="container-fluid"
>
  <div
    className="row cars-overview-grid"
  >
    <div
      className="col-md-12"
    >
      <div
        className="loading-spinner "
      >
        <div
          className="sk-circle"
        >
          <div
            className="sk-circle1 sk-child"
          />
          <div
            className="sk-circle2 sk-child"
          />
          <div
            className="sk-circle3 sk-child"
          />
          <div
            className="sk-circle4 sk-child"
          />
          <div
            className="sk-circle5 sk-child"
          />
          <div
            className="sk-circle6 sk-child"
          />
          <div
            className="sk-circle7 sk-child"
          />
          <div
            className="sk-circle8 sk-child"
          />
          <div
            className="sk-circle9 sk-child"
          />
          <div
            className="sk-circle10 sk-child"
          />
          <div
            className="sk-circle11 sk-child"
          />
          <div
            className="sk-circle12 sk-child"
          />
        </div>
        <div
          className="loading-spinner__text"
        >
          <span>
            Loading data, please wait...
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
